File: backend/gameHubBack/services/BatalhaRural/Rules/CollisionRuler.cs
using gameHubBack.Enums.BatalhaRural;

namespace gameHubBack.Services.BatalhaRural.Rules;

public static class CollisionRuler
{
    public static bool IsTileOccupied(TokenInfo tokenInfo, string[][] tiles)
    {
        var tileStatus = tiles[tokenInfo.BaseZeroPosY][tokenInfo.BaseZeroPosX];

        return tileStatus != "O";
    }

    public static bool IsTokenHorizontallyOverlapping(TokenInfo tokenInfo, string[][] tiles, Direction direction)
    {
        if (direction == Direction.Right)
        {
            if (IsTokenDiagonal(tokenInfo, Axis.X)) return true;

            for (var i = tokenInfo.BaseZeroPosX; i < tokenInfo.BaseZeroPosX + tokenInfo.TokenSize; i++)
            {
                if (tiles[tokenInfo.BaseZeroPosY][i] != "O") return true;
            }
        }

        if (direction == Direction.Left)
        {
            if (IsTokenDiagonal(tokenInfo, Axis.X)) return false;

            for (var i = 0; i < tokenInfo.TokenSize; i++)
            {
                if (tiles[tokenInfo.BaseZeroPosY][tokenInfo.BaseZeroPosX - i] != "O") return true;
            }
        }

        return false;
    }

    public static bool IsTokenVerticallyOverlapping(TokenInfo tokenInfo, string[][] tiles, Direction direction)
    {
        if (direction == Direction.Up)
        {
            if (IsTokenDiagonal(tokenInfo, Axis.Y)) return false;

            for (var i = 0; i < tokenInfo.TokenSize; i++)
            {
                if (tiles[tokenInfo.BaseZeroPosY - i][tokenInfo.BaseZeroPosX] != "O") return true;
            }
        }

        if (direction == Direction.Down)
        {
            if (IsTokenDiagonal(tokenInfo, Axis.Y)) return false;

            for (var i = tokenInfo.BaseZeroPosY; i < tokenInfo.BaseZeroPosY + tokenInfo.TokenSize; i++)
            {
                if (tiles[i][tokenInfo.BaseZeroPosX] != "O") return true;
            }
        }

        return false;
    }

    private static bool IsTokenDiagonal(TokenInfo tokenInfo, Axis axis)
    {
        return axis switch
        {
            Axis.X => tokenInfo.BaseZeroFinalPosY != tokenInfo.BaseZeroPosY,
            Axis.Y => tokenInfo.BaseZeroFinalPosX != tokenInfo.BaseZeroPosX,
            _ => false
        };
    }
}

File: frontend/GameHubFront/src/Games/batalhaRural/services/rules/CollisionRuler.ts
import { Board } from "../../entities/Board";
import { Axis } from "../../enums/Axis";
import { Direction } from "../../enums/Direction";
import { Logger } from "../../helpers/Logger";
import { TokenInfo } from "../../helpers/TokenInfo";

export class CollisionRuler {

    static isTileOccupied(tokenInfo: TokenInfo, board: Board): boolean {
        const tileStatus = board.tiles[tokenInfo.baseZeroPosY][tokenInfo.baseZeroPosX];

        if (tileStatus != 'O') { Logger.collision(); return true; }

        return false;
    }

    static isTokenHorizontallyOverlaping(tokenInfo: TokenInfo, board: Board, direction: Direction): boolean {
        if (direction == Direction.Right) {
            if (this.isTokenDiagonal(tokenInfo, Axis.X)) return true;

            for (let i = tokenInfo.baseZeroPosX; i < tokenInfo.baseZeroPosX + tokenInfo.tokenSize; i++) {
                let char = board.tiles[tokenInfo.baseZeroPosY][i];

                if (char != 'O') { Logger.collision(); return true; }
            }
        }

        // verificando se há parte de uma peça em algum ponto da posição desejada
        if (direction == Direction.Left) {
            if (this.isTokenDiagonal(tokenInfo, Axis.X)) return false;

            for (let i = 0; i < tokenInfo.tokenSize; i++) {
                let char = board.tiles[tokenInfo.baseZeroPosY][tokenInfo.baseZeroPosX - i];

                if (char != 'O') { Logger.collision(); return true; }
            }
        }

        return false;
    }

    static isTokenVerticallyOverlaping(tokenInfo: TokenInfo, board: Board, direction: Direction): boolean {

        // verificando se há parte de uma peça em algum ponto da posição desejada
        if (direction == Direction.Up) {
            if (this.isTokenDiagonal(tokenInfo, Axis.Y)) return false;

            for (let i = 0; i < tokenInfo.tokenSize; i++) {
                let char = board.tiles[tokenInfo.baseZeroPosY - i][tokenInfo.baseZeroPosX];

                if (char != 'O') { Logger.collision(); return true; }
            }
        }

        // verificando se há parte de uma peça em algum ponto da posição desejada
        if (direction == Direction.Down) {
            if (this.isTokenDiagonal(tokenInfo, Axis.Y)) return false;

            for (let i = tokenInfo.baseZeroPosY; i < tokenInfo.baseZeroPosY + tokenInfo.tokenSize; i++) {
                let char = board.tiles[i][tokenInfo.baseZeroPosX];

                if (char != 'O') { Logger.collision(); return true; }
            }
        }

        return false;
    }

    private static isTokenDiagonal(tokenInfo: TokenInfo, axis: Axis): boolean {
        switch (axis) {
            case Axis.X:
                if (tokenInfo.baseZeroFinalPosY != tokenInfo.baseZeroPosY) { Logger.diagonal(); return true; }
                break;
            case Axis.Y:
                if (tokenInfo.baseZeroFinalPosX != tokenInfo.baseZeroPosX) { Logger.diagonal(); return true; }
                break;
        }

        return false;
    }
}